### PR TITLE
Use adapted time step from sub-app for loose coupling

### LINKED ISSUE #7842: Behavior of IterationAdaptiveTimeStepper as sub app
### Description of the enhancement or error report

The behavior occurs when `IterationAdaptiveDT` is used in the sub app, without sub-cycling and 
the master does not constrain the time step. In that case if the time step should be increased 
due to fast convergence, it does not happen until one time step later. 
### Rationale for the enhancement or information for reproducing the error

An example will be provided. 
### Identified impact

(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

### PR DESCRIPTION
**Pull request by lindsayad**
- Increment sub application time steps before `dt` is selected in master application, such that the `dt` selection is made with correct updated information
- Split the `MultiApp` `advanceStep` method into two methods, one responsible for incrementing the time step, and another responsible for doing post step activities (`Executioner::postStep` and `Executioner::endStep`)
- Add some documentation because I think I can say our transient systems are a little bewildering

Closes #7842

### CODE REVIEW

**Comment by lindsayad** [comment]
If we decide we're ok with the small interface changes, I'll create PRs for the failing applications

File: framework/include/multiapps/MultiApp.h
Comment (permcody) [inline framework/include/multiapps/MultiApp.h]: Hmm, Are you sure about this statement? We do have a few Picard tests in the suite and this PR just passed them with recover.  Note that the master writes all of the recovery data for the simulation so as long as things are in the same state they were written in, things generally work out. If you still expect it to fail, we need to dive in and look at this more carefully.

File: framework/include/multiapps/MultiApp.h
Comment (permcody) [inline framework/include/multiapps/MultiApp.h]: This is a pretty significant API update (since these are pure virtuals). You should probably leave the old one there which just calls the new one but also use `mooseDeprecated`. This will mean that the new API can't be pure virtual for now, but it looks like Rattlesnake is failing due to this update so you don't really have a good option. You'll just have to come back through after apps are updated and change this.

File: framework/src/base/FEProblemBase.C
Comment (permcody) [inline framework/src/base/FEProblemBase.C]: I'm not sure if I'm crazy about these new messages. We have two capitalized words followed by this cryptic "t_step" thing. Can you spell it out, or just leave it off like it was before? The other issue with "t_step" is that sometimes the multiapps take several steps and sometimes they take one.

**Review by permcody — DISMISSED** [review]
Looks pretty good, just a few minor comments.

File: framework/src/multiapps/TransientMultiApp.C
Comment (andrsd) [inline framework/src/multiapps/TransientMultiApp.C]: Remove the comments, we do not do that. A decent editor can take you to the matching `{` and back.

File: test/tests/multiapps/loose_couple_time_adapt/gold/begin.e
Comment (andrsd) [inline test/tests/multiapps/loose_couple_time_adapt/gold/begin.e]: This is interesting. Are Exodus files not binary?

File: test/tests/multiapps/loose_couple_time_adapt/adaptiveDT.i
Comment (andrsd) [inline test/tests/multiapps/loose_couple_time_adapt/adaptiveDT.i]: Get rid of the active lines, they are not needed. And in the other input files as well.

**Review by andrsd — CHANGES_REQUESTED** [review]
I am fine with the changes, but I think @friedmud should look at this as well, since it is mostly his design.

File: test/tests/multiapps/loose_couple_time_adapt/gold/begin.e
Comment (lindsayad) [reply test/tests/multiapps/loose_couple_time_adapt/gold/begin.e]: This is actually a symlink, since both tests should yield the exact same exodus results. This works fine, but if we want real exodus files for both tests I can do that.

Comment (andrsd) [reply test/tests/multiapps/loose_couple_time_adapt/gold/begin.e]: Symlinks are fine. I just was not able to tell it was a symlink from the diff 👍

File: framework/include/multiapps/MultiApp.h
Comment (lindsayad) [reply framework/include/multiapps/MultiApp.h]: Yep, turns out Picard does work, although for a bit of a convoluted reason...The time of the sub-app *is* written out behind the time of the master, e.g. for a 20 step simulation with `dt = 0.1` and running `--half-transient`, the restartable data has the master's time as 1, and the stored sub-apps' time as 0.9. However, upon recover, the sub-app's time is caught-up in `Transient::incrementStepOrReject` because `_picard_max_its > 1` before any solves are performed.

**Review by andrsd — DISMISSED** [review]

**Comment by lindsayad** [comment]
I don't recall dismissing @andrsd review! But anyways, I think this is ready to go...

**Comment by andrsd** [comment]
> I don't recall dismissing @andrsd review!

That's probably because you pushed and modified lines I commented on. That dismissed the review.

**Comment by lindsayad** [comment]
ping @permcody

**Comment by permcody** [comment]
This fails Tiamat, not surprising since it uses this logic heavily. We'll have to look at these failures and probably help the Tiamat maintainers regold.

**Comment by clarnokt** [comment]
what can we do to help?

**Comment by permcody** [comment]
@clarnokt - Two options and they aren't independent.

1) Grant @lindsayad access to the Tiamat repo so he can take a look at the changes and supply a patch. This may not require any real changes to Tiamat, but it is correcting behavior so your gold files will have to be re-verified

2) Pull these changes on your side and see if you can make sense of what's changing. Alex is available to help in either case.

**Comment by clarnokt** [comment]
@lindsayad has been invited to join the developer group.
@stimpsonsg or I will try to take a look tomorrow.

**Comment by lindsayad** [comment]
Joined. I'll try to take a look at this either today or tomorrow.

**Comment by permcody** [comment]
30 day limit exceeded on this PR, but it's important. We need to give it a kick and see if we can resolve issues with Tiamat or whatever else is going on here.

**Comment by permcody** [comment]
@friedmud - Can you review this PR?

**Comment by friedmud** [comment]
On it

**Comment by friedmud** [comment]
Ok - I see what this is doing now.  Makes sense.

Once all of the tests pass I will merge.

**Review by friedmud — APPROVED** [review]
👍

**Comment by permcody** [comment]
This PR causes a Tiamat failure. One of the tests isn't generating files anymore, the other is an Exodiff.  We will have to work with @stimpsonsg to regold the one if the new solution is correct. The other test is not a difference test so we should be able to figure it out here.

**Comment by lindsayad** [comment]
From the time I spent looking at this a bit last week, the failures look
real, i.e. the results produced after this PR are not correct

**Comment by lindsayad** [comment]
@stimpsonsg @clarnokt It would be really helpful for debugging issues with Tiamat if we could run in debug mode. However, when I do I get malloc assertion errors. See the following backtrace. Any illumination this?
```
#0  __GI_abort () at abort.c:51
#1  0x00007fb61f5042e8 in __malloc_assert (
    assertion=assertion@entry=0x7fb61f618190 "(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)", 
    file=file@entry=0x7fb61f614bc5 "malloc.c", line=line@entry=2394, 
    function=function@entry=0x7fb61f6189d8 <__func__.11509> "sysmalloc") at malloc.c:301
#2  0x00007fb61f508426 in sysmalloc (nb=nb@entry=80, av=av@entry=0x7fb61f84bb20 <main_arena>) at malloc.c:2391
#3  0x00007fb61f509743 in _int_malloc (av=av@entry=0x7fb61f84bb20 <main_arena>, bytes=bytes@entry=64)
    at malloc.c:3827
#4  0x00007fb61f50b184 in __GI___libc_malloc (bytes=bytes@entry=64) at malloc.c:2913
#5  0x00007fb6209bf918 in operator new (sz=64) at ../../../../gcc-6.2.0/libstdc++-v3/libsupc++/new_op.cc:50
#6  0x000000000047cb8c in __gnu_cxx::new_allocator<int*>::allocate (this=0x7ffd2d4f329f, __n=8)
    at /opt/moose/gcc-6.2.0/include/c++/6.2.0/ext/new_allocator.h:104
#7  0x000000000047bf8a in std::allocator_traits<std::allocator<int*> >::allocate (__a=..., __n=8)
    at /opt/moose/gcc-6.2.0/include/c++/6.2.0/bits/alloc_traits.h:416
#8  0x000000000047ab3f in std::__cxx1998::_Deque_base<int, std::allocator<int> >::_M_allocate_map (
    this=0x26eee80, __n=8) at /opt/moose/gcc-6.2.0/include/c++/6.2.0/bits/stl_deque.h:614
#9  0x0000000000479011 in std::__cxx1998::_Deque_base<int, std::allocator<int> >::_M_initialize_map (
    this=0x26eee80, __num_elements=0) at /opt/moose/gcc-6.2.0/include/c++/6.2.0/bits/stl_deque.h:687
#10 0x000000000047689a in std::__cxx1998::_Deque_base<int, std::allocator<int> >::_Deque_base (this=0x26eee80)
    at /opt/moose/gcc-6.2.0/include/c++/6.2.0/bits/stl_deque.h:490
#11 0x00000000004735b4 in std::__cxx1998::deque<int, std::allocator<int> >::deque (this=0x26eee80)
    at /opt/moose/gcc-6.2.0/include/c++/6.2.0/bits/stl_deque.h:884
#12 0x000000000046ea9d in std::__debug::deque<int, std::allocator<int> >::deque (this=0x26eee68)
    at /opt/moose/gcc-6.2.0/include/c++/6.2.0/debug/deque:87
#13 0x000000000046ebf3 in Teuchos::basic_FancyOStream_buf<char, std::char_traits<char> >::basic_FancyOStream_buf (
    this=0x26eed98, oStream=..., tabIndentStr=..., startingTab=0, showLinePrefix=false, maxLenLinePrefix=10, 
    showTabCount=false, showProcRank=false)
    at /opt/moose/trilinos-release-12-6-2/mpich-gcc-opt/include/Teuchos_FancyOStream.hpp:873
#14 0x000000000046a313 in Teuchos::basic_FancyOStream<char, std::char_traits<char> >::basic_FancyOStream (
    this=0x26eed90, oStream=..., tabIndentStr=..., startingTab=0, showLinePrefix=false, maxLenLinePrefix=10, 
    showTabCount=false, showProcRank=false, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /opt/moose/trilinos-release-12-6-2/mpich-gcc-opt/include/Teuchos_FancyOStream.hpp:1261
#15 0x00007fb62b0c3dce in Teuchos::fancyOStream (startingTab=0, showLinePrefix=false, maxLenLinePrefix=10, 
    showTabCount=false, showProcRank=false, tabIndentStr=..., oStream=...)
    at /tmp/moose_package_build_temp/trilinos-mpich-gcc-opt/Trilinos-trilinos-release-12-6-2/packages/teuchos/core/src/Teuchos_FancyOStream.hpp:602
#16 pike::MultiphysicsDistributor::buildOStreams (this=this@entry=0x26de1a0)
    at /tmp/moose_package_build_temp/trilinos-mpich-gcc-opt/Trilinos-trilinos-release-12-6-2/packages/pike/blackbox/src/Pike_MultiphysicsDistributor.cpp:236
#17 0x00007fb62b0c5c44 in pike::MultiphysicsDistributor::setup (this=0x26de1a0, globalComm=..., 
    printCommDistribution=<optimized out>)
    at /tmp/moose_package_build_temp/trilinos-mpich-gcc-opt/Trilinos-trilinos-release-12-6-2/packages/pike/blackbox/src/Pike_MultiphysicsDistributor.cpp:170
#18 0x000000000045d618 in main (argc=9, argv=0x7ffd2d4f56c8)
    at /home/lindad/projects-trilinos/Tiamat/src/main.C:123
```

**Comment by stimpsonsg** [comment]
I can't say that I've seen that one before.  And it does appear to be a Trilinos issue...agreed.

**Comment by brianmoose** [comment]
The test `multiapps/loose_couple_time_adapt.end` seems to reliably exodiff on on the SLEPc target. See #10705 
It also seems to sporadically exodiff on the Trilinos target #10706, Min clang target #10708, and the Boost target #10707

**Comment by lindsayad** [comment]
So @stimpsonsg you've been able to run tiamat before in debug mode?

**Comment by stimpsonsg** [comment]
My experience running Mock Tiamat has been somewhat limited, so I haven't.  Last time I did try in debug, I also hit a segfault, but it didn't seemTrilinos-related:

```
Tiamat_MultiApp2: Time Step  0, time = 0
Tiamat_MultiApp2:                 dt = 0
Tiamat_MultiApp2: 
Tiamat_MultiApp2: Postprocessor Values:
Tiamat_MultiApp2: +----------------+-----------------------+--------------------------+
Tiamat_MultiApp2: | time           | bison_pin_total_power | casl_max_rod_temperature |
Tiamat_MultiApp2: +----------------+-----------------------+--------------------------+
Tiamat_MultiApp2: |   0.000000e+00 |          0.000000e+00 |             0.000000e+00 |
Tiamat_MultiApp2: +----------------+-----------------------+--------------------------+
Tiamat_MultiApp2: 
[2]PETSC ERROR: ------------------------------------------------------------------------
[2]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[2]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[1]PETSC ERROR: [2]PETSC ERROR: or see http://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
------------------------------------------------------------------------
```

at what point in the calculation are you seeing it error out?

**Comment by lindsayad** [comment]
We don't even get to the MOOSE header:
```
lindad@cone:~/projects-trilinos/Tiamat/tests/subcycle(devel)$ mpiexec -n 4 ../../tiamat-dbg --n-threads=1 -i dummy.i --num-mpact-procs=2 --num-bison-apps=3 --no-gdb-backtrace

../../tiamat-dbg : Warning, the 1st option '--n-threads=1' was not recognized (use --help)!

../../tiamat-dbg : Warning, the 2nd option '-i' was not recognized (use --help)!

../../tiamat-dbg : Warning, the 3rd option 'dummy.i' was not recognized (use --help)!

../../tiamat-dbg : Warning, the 6th option '--no-gdb-backtrace' was not recognized (use --help)!
tiamat-dbg: malloc.c:2394: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
tiamat-dbg: malloc.c:2394: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
tiamat-dbg: malloc.c:2394: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
tiamat-dbg: malloc.c:2394: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 13890 RUNNING AT cone
=   EXIT CODE: 134
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Aborted (signal 6)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

**Comment by lindsayad** [comment]
@stimpsonsg Didn't resolve running in debug, but the tests appear to be running again after the merge request in CASL/Tiamat#18

**Pull request merged**
